### DISCUSSION TITLE
Cannot select from auth.users inside a function called by a trigger

### DISCUSSION THREAD

**Discussion by joaoguidev** [supabase/supabase#21571] [Questions] [answered]
Hello.

Everytime a user is added to my auth.users table, depending on their role I want to insert data in different tables. For instance user has custom_role_A save id to table public.table_A and custom_role_B save id to table_B. The problem is that I cannot define the user_role inside the function that the trigger calls and I don't know why.

Here is the **trigger**:

```
create trigger do_something
after insert on auth.users for each row
execute procedure public.do_something();
```

And here is the **function**.On the RAISE LOG, user_id variable is coming back null and new.id logs the user id as expected:

```
declare
  user_role TEXT;
begin
  select role into user_role from auth.users where id = new.id;
  RAISE LOG 'User Role: %', user_role;
  RAISE LOG 'New Id: %', new.id;
  if user_role = 'custom_role_A' then
    insert into public.table_A(id, email)
    values(new.id, new.email);
  elsif user_role = 'custom_role_B' then
    insert into public.table_B(id)
    values (new.id);
  end if;
  return new;
end;
```

The security of this function is **definer**.

 **Things that I tried:**

- Instead of using the variable user_role I manually pasted an id in` "select role into user_role from auth.users where id = wsa45a5sd4q5w"`. It works but not with the variable user_id that i declared. 
- Define user_role to UUID instead of TEXT but than it throws an error on the postgres log and cancel the transaction. 
- Cast new.id to new.id::text but I get the error "No operator matches the given name and argument types. You might need to add explicit type casts." on postgres log and the transaction is aborted

I assumed that because the trigger is set to **after insert**  the new entry in the auth.user table would be available to be selected.

**Comment by GaryAustin1** [comment]
Why are you not just using new.role instead of the select?

**Reply by joaoguidev** [reply]
Hello Gary. Thank you for replying.

The role is not coming on the new object. new.role comes null, maybe that might be a bug. Here is the RAISE LOG of the new object:

Output of  `RAISE LOG 'New here: %', new;` 
(00000000-0000-0000-0000-000000000000,s32fsdf2-3sd123-vsef12-bwer2323-fsdfRW32R2,authenticated,"",textinghereabc@gmail.com,$e1$43$f.@sdf23rFAEFsd/23fdsfewf.dsad123er,,,"",,"",,"","",,,"{""provider"": ""email"", ""providers"": [""email""]}",{},,"2024-02-26 18:41:25.673547+00","2024-02-26 18:41:25.673547+00",,,"","",,,"",0,,"",,f,)

The "authenticated" above refers to new.aud and **not** new.role.

output of  `RAISE LOG 'New aud: %', new.aud;`
![image](https://github.com/supabase/supabase/assets/63625334/4d36e47a-c5a6-4110-a8d2-e525c82dd2c9)



Output of new.role is null
Output of  `RAISE LOG 'New Role: %', new.role;` is null 

![image](https://github.com/supabase/supabase/assets/63625334/cba1b55b-8696-4a62-bf6b-b187796e440e)

I have created the user like this:

```
   const { data, error } = await supabaseAdmin.auth.admin.createUser({
      email: "textinghereabc@gmail.com",
      password: "Testing@123",
      role: "custom_role_A",
   })
```
I was expecting the new.role would be 'custom_role_A'

I have the impression that instead of the new come with the **role** it is coming with the **aud** instead.

**Reply by GaryAustin1** [reply]
It could be they do the role in the update request that follows the insert.  I can’t confirm you can set the role thru admin update. But the role should be authenticated if not your role if not.  This says to me it is on the update that will follow.  You can confirm by looking at the role column after user is created.

**Reply by joaoguidev** [reply]
In auth.users the new user role column gets in there correctly, "custom_role_A" in the column role and "authenticated" in the column aud.
 There was a discussion about setting custom roles here: [https://github.com/orgs/supabase/discussions/11948](https://github.com/orgs/supabase/discussions/11948)

And on the documentation role is one of the parameters accepted on admi.createUser()
![image](https://github.com/supabase/supabase/assets/63625334/c799dca0-3f5e-4fc5-8aee-7d864907e413)

Regarding acessing the role inside the function it seems that it is not possible or would you suggest a way of doing that?

I tried with the select as shown on the beginingg of this discussion and it also did not work either. I tried this because the trigger that triggered this function is "after insert" therefore I was hoping to be able to get the role with the select inside the function. 

The best would be to be able to select the role just by using new.role and role beeing in the new object. But for some reason it is not. Instead what is in the new object is new.aud as shown on the previous post.

**Reply by GaryAustin1** [reply]
You will need to add an update trigger and function.  The update comes right after the insert.

**Reply by joaoguidev** [reply]
But if I do the trigger on update on auth.users it will be triggered not only when the user is inserted but also, lets say, when a existing user updates their email or password.

**Reply by GaryAustin1** [reply, marked as answer]
Yes. You would need to add when clause on new versus old role changing to reduce the calls.  The only way you will get the role in a trigger is on update as it is pretty clear they don't set it on the insert.
